core/types/calls.cc

```diff
@@ -2615,6 +2615,110 @@ class Shape_to_hash : public IntrinsicMethod {
     }
 } Shape_to_hash;
 
+// `<Magic>.<to-hash-dup>(x)` and `<Magic>.<to-hash-nodup>(x) both behave like `x.to_hash`
+class Magic_toHash : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        ENFORCE(args.args.size() == 1);
+
+        auto &arg = args.args[0];
+        InlinedVector<LocOffsets, 2> argLocs;
+        CallLocs locs{args.locs.file, args.locs.call, args.locs.call, argLocs};
+        InlinedVector<const TypeAndOrigins *, 2> innerArgs;
+
+        DispatchArgs dispatch{core::Names::toHash(),
+                              locs,
+                              0,
+                              innerArgs,
+                              arg->type,
+                              {arg->type, args.fullType.origins},
+                              arg->type,
+                              nullptr,
+                              args.originForUninitialized};
+        res = arg->type.dispatchCall(gs, dispatch);
```

Comment:
Do you know if we would have tests that ensure something like `foo({**x, a: 1})` (i.e. the types derived by this function and others) actually do what we think they do?  I guess that example might run afoul of Ruby's hash-args-are-kwargs behavior, but maybe there are other ways we could test this?  (Straight `T.reveal_type` tests?)

Reply:
The `test/testdata/desugar/hash.rb` test covers a lot of cases like this. I can look at adding any cases when reworking it to better reflect the new desugaring.

Reply:
Actually this case looks like it's already present in https://github.com/sorbet/sorbet/blob/35c488b6/test/testdata/infer/kwargs_splat.rb#L25

Reply:
That makes sense, it seemed like there ought to be some cases for kwsplat typing somewhere. :)